### PR TITLE
grouping functions

### PR DESCRIPTION
**Pull request by elizavetaRa**
## Description
Implementation of grouping functions for the table

## Related Issue
https://its.cern.ch/jira/secure/RapidBoard.jspa?rapidView=6014&projectKey=CERNBOX&view=detail&selectedIssue=CERNBOX-1901&quickFilter=11535

https://its.cern.ch/jira/secure/RapidBoard.jspa?rapidView=6014&projectKey=CERNBOX&view=detail&selectedIssue=CERNBOX-1902&quickFilter=11535

https://its.cern.ch/jira/secure/RapidBoard.jspa?rapidView=6014&projectKey=CERNBOX&view=detail&selectedIssue=CERNBOX-1980&quickFilter=11535

## Motivation and Context
Usability improvement

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests

## Checklist:

- [x] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation added/updated

## Open tasks:
- Commenting
- Sorting + regrouping on th click?

### CODE REVIEW

File: src/components/table/OcTableFiles.vue [hunk 1]
```diff
@@ -711,64 +711,210 @@ export default {
...
-            thumbnail: "https://cdn.pixabay.com/photo/2015/09/09/16/05/forest-931706_960_720.jpg",
-            indicators: [],
-            type: "file",
-            sdate: "Mon, 11 Jan 2021 14:34:04 GMT",
-            owner: [{
+          groupingSettings(){
+      return {
+        groupingAllowed: true,
+        defaultGroupingBy: "None",
+        passedGroupingBy: "",  //None, creation, owner,alphabetically
```
Comment (diocas) [inline src/components/table/OcTableFiles.vue:718]: Why do we still need this? Shouldn't we use the keys of `groupingFunctions`?

File: src/components/table/OcTableFiles.vue [hunk 2]
```diff
@@ -711,64 +711,210 @@ export default {
...
+        showGroupingOptions: true,
```
Comment (diocas) [inline src/components/table/OcTableFiles.vue:719]: What does this option do? Can't we just check if we have grouping functions to enable/disable the functionality?

File: src/components/table/OcTableFiles.vue [hunk 3]
```diff
@@ -711,64 +711,210 @@ export default {
...
+        previewTable: true,
```
Comment (diocas) [inline src/components/table/OcTableFiles.vue:720]: Same for here: can't we use `previewAmount` only? If it's 0, is off, it's it's not 0, then we take it.

File: src/components/table/OcTable.vue [hunk 4]
```diff
@@ -216,8 +399,89 @@ export default {
...
+        result.forEach(e => {
+          let eLength = e.data.length
+          if (eLength - previewCount >= 0) e.data = e.data.slice(0, previewCount - 1)
+          previewCount = previewCount - eLength
+        })
+
+        result = result.filter(e => e.data.length > 0)
+      }
+
+      this.resultArray = result
```
Comment (diocas) [inline src/components/table/OcTable.vue:422]: Am I understanding correctly that you store this in the state to allow you to hide/show the rows, right?
This seems a bit messy, but I don't know enough vue to know if it's necessary or not. I guess I'll let oc review this after.

File: src/components/table/OcTable.vue [hunk 5]
```diff
@@ -216,8 +399,89 @@ export default {
...
+            ? result[this.groupingSettings.groupingFunctions[col](row)].push(row)
+            : (result[this.groupingSettings.groupingFunctions[col](row)] = [row])
+        })
+        return this.produceGroups(Object.entries(result), asc, col)
+      }
+    },
+
+    produceGroups(arr, asc, col) {
+      let resultArray = []
+      for (const [key, value] of arr) {
```
Comment (diocas) [inline src/components/table/OcTable.vue:450]: Why did you create a dict before, just to pass it here and create another dict? Why don't we merge this functionality in the calling function instead?

**Review by diocas — CHANGES_REQUESTED** [review]
It's hard to follow when there are no comments describing what is being done, decisions, etc.
I don't know if oc is documenting their stuff, but at least we should.

Just some comments and I think we can merge. I'll let the proper review be done by oc.

File: src/components/table/OcTableFiles.vue (hunk 1 above)
Comment (elizavetaRa) [reply src/components/table/OcTableFiles.vue:718]: - it's no possible to use groupingFunctions (for example first key) for default grouping because the order of enumeration is not garanteed for the objects.
- passedGroupingBy is needed if you want to be able to change which grouping to use from outside of the component

File: src/components/table/OcTableFiles.vue (hunk 2 above)
Comment (elizavetaRa) [reply src/components/table/OcTableFiles.vue:719]: showGroupingOptions is needed to enable/disable oc-select element with the listed grouping options inside the component. You may want to create your own selection element from outside of the table component (what we discussed), in that case you don't pass the prop or pass "false"

File: src/components/table/OcTableFiles.vue (hunk 3 above)
Comment (elizavetaRa) [reply src/components/table/OcTableFiles.vue:720]: good point

File: src/components/table/OcTableFiles.vue (hunk 1 above)
Comment (diocas) [reply src/components/table/OcTableFiles.vue]: Yes, I did not mention `defaultGroupingBy`, only `passedGroupingBy`. Do we really need to declare it here for being able to use it externally? Can't we define it somewhere else internally?
Btw, how do you declare an external grouping dropdown? Can you give me an example to understand?

File: src/components/table/OcTableFiles.vue (hunk 2 above)
Comment (diocas) [reply src/components/table/OcTableFiles.vue]: Understood!

File: src/components/table/OcTable.vue (hunk 5 above)
Comment (diocas) [reply src/components/table/OcTable.vue]: Ups, I thought I had deleted this comment.. After re-reading I understood the reason and what I said was not true.

File: src/components/table/OcTable.vue [hunk 6]
```diff
@@ -1,68 +1,210 @@
...
+               {
+                text-align: center;
+              }
+            "
+            ><div class="preview-heading">Show more</div> </oc-td
+          ><oc-td
+            v-else
+            key="showLess"
+            :colspan="fields.length"
+            style="
```
Comment (diocas) [inline src/components/table/OcTable.vue:188]: One thing I forgot to mention before: can't we put the styles in the styles file instead of inline? I might not want this to be centered.

**Comment by diocas** [comment]
I'm merging this PR but please address my comments in a new PR.
